src/python/pants/base/specs.py

```diff
@@ -319,22 +319,58 @@ def __iter__(self) -> Iterator[AddressSpec]:
     return iter(self.dependencies)
 
 
+class FilesystemSpec(Spec, metaclass=ABCMeta):
+  pass
+
+
 @dataclass(frozen=True)
-class FilesystemSpec(Spec):
-  """The glob may be anything understood by `PathGlobs`.
+class FilesystemLiteralSpec(FilesystemSpec):
+  """A literal file name, e.g. `foo.py`."""
+  file: str
 
-  The syntax supported is roughly Git's glob syntax."""
+  def to_spec_string(self) -> str:
+    return self.file
+
+
+@dataclass(frozen=True)
+class FilesystemGlobSpec(FilesystemSpec):
+  """A spec with a glob or globs, e.g. `*.py` and `**/*.java`."""
   glob: str
 
   def to_spec_string(self) -> str:
     return self.glob
 
 
+@dataclass(frozen=True)
+class FilesystemIgnoreSpec(FilesystemSpec):
+  """A spec to ignore certain files or globs."""
+  glob: str
+
+  def __post_init__(self) -> None:
+    if self.glob.startswith("!"):
+      raise ValueError(f"The `glob` for {self} should not start with `!`.")
```

Comment:
A parser generator, or at least some central point that converts string specs => `Spec` class instances, could help to avoid one-off, possibly error-prone double-checking in the constructor here (because we could rely on the parser to create correct specs instead of having the logic spread over many classes).

Again, *not* necessary right now, but it seems like a very fruitful idea if/when it seems like we're outgrowing this approach. I also plan to do a similar thing with parsing args for `--query`.